### DISCUSSION TITLE
need help with fixing russian language

### DISCUSSION THREAD

**Discussion by kirill-jjj** [espeak-ng/espeak-ng#1427] [Q&A] [answered]
Hi, I need some help here. We have words in Russian with two stress's, for example (аудиокнига), (аудиоплеер), (видеоплеер), and so on. What is the best way to correct such words?
I corrected the word аудиокнига, but I don't like how I did it. Just making a transcription of a word in ru_list will not solve all the problems, because there are still such forms of words as аудиокниги, аудиокнигой

**Comment by jaacoppi** [comment]
What is the correct phoneme output for these words? What kid of stress, two primary stresses, or one primary and one secondary? Which syllables should be stressed? Why do these words have two stresses and other Russian words don't?

in current master:
 $ espeak-ng -v ru "аудиокнига" -x
'Aud;iVkn;iga#

What should it be?

**Reply by kirill-jjj** [reply]
I never got used to pressing the reply button. I made a pull request yesterday and now it should be like 'Aud;iV#kn;'iga#

**Reply by kirill-jjj** [reply]
I think it should be like this, the letter а has a secondary stress, and the letter и, the primary, But I need to hear how it sounds.

**Comment by kirill-jjj** [comment]
that is strange, I made the pull request yesterday, and it should be like 'Aud;iV#kn;'iga#. this word consists of two parts, аудио, and книга, which книга? аудио. It's just that it's usually written together, like аудиокнига , but if you separate the parts with spaces, you will get the correct pronunciation, аудио книга.

**Comment by jaacoppi** [comment, marked as answer]
Sorry, missed your PR in master.

You're talking about compound words. They are very tricky. There's no code to support them. Some languages use rules to match them but it's impossible to match all combinations.

Do a search for "compound" in dictsource/*_rules to get some ideas.

**Reply by kirill-jjj** [reply]
Thank you, but can I still wright them in ru_list? or will ru_listx be better?

**Comment by jaacoppi** [comment]
You can use list files, but rules are more flexible in most cases.

Take a look at line counts for listx files. Ru_listx is the biggest file with 202194 entries. Next comes Mandarin with only 77864. This either means that russian grammar is very complicated and unsuitable for espeak NG rules format OR the words shouldn be in the rules file.

Looking at the contents of ru_listx it seems that it's only used for setting primary stress. language option stressRule and ru_rules should be used for that instead.

I think using list or listx makes no difference on modern machines. It was important in the 1990s with limited hardware.

**Reply by kirill-jjj** [reply]
by the way, I extended ru_listx, it was the size 199805, can I continue to do this?

**Reply by kirill-jjj** [reply]
and no, ru_listx is needed, here writn whi
https://espeak.sourceforge.net/data/

**Reply by kirill-jjj** [reply]
and my friend and I fixed the digit system for Russian, it was wrong and strange, it didn't know how to speak some numbers at all, for example 1000000 and 1000000000, some digits were not sed correctly, for example 2000000, 2000, 3000000, 3000, and so on.

**Reply by kirill-jjj** [reply]
look at this, the number 1000, he read almost correctly, odn'A t'ys;VtS;, although the form of the numeral for such a case is not correct, but do you know how he read the number 1000000 and 1000000000 before our corrections? ojd;'in t'ys;VtS;! I think you will understand without me that this number should not be like a thousand. And this bug remained in the Russian language for almost 8 years

**Comment by jaacoppi** [comment]
It looks like stress rules in Russian are difficult:
https://russianstress.info/index.php/main/view/stress
https://blogs.transparent.com/russian/word-stress-patterns-in-russian/

Using list instead of rules for stress makes sense.

Russian currently uses stressRule STRESSPOSN_SYLCOUNT. That could be improved as well.

**Reply by kirill-jjj** [reply]
there is no better algorithm for Russian than here

**Reply by kirill-jjj** [reply]
than this, sorry

**Reply by kirill-jjj** [reply]
using translater

**Comment by kirill-jjj** [comment]
but if it were possible to set a piece of a word with some kind of flag, of course without removing the ability to write words it entirety, I think it would simplify the process of adding words to ru_listx, like this, аудиокни*     $5

**Reply by kirill-jjj** [reply]
as for compound words, I'll try to figure out how to add them, maybe through ru_rules, maybe through ru_listx

**Reply by kirill-jjj** [reply]
thank you for giving me the possibility to fix Russian language in espeak

**Comment by jaacoppi** [comment]
> but if it were possible to set a piece of a word with some kind of flag, of course without removing the ability to write words it entirety

This works:
_) аудиокни aud;iVkn;'I

I'm not sure about flags in _rules files. I think they won't work.

**Reply by kirill-jjj** [reply]
got it, by the way, have you seen my issue about android espeak?

**Comment by jaacoppi** [comment]
> got it, by the way, have you seen my issue about android espeak?

Unfortunately I know absolutely nothing about android espeak. Can't help you.
